File: code/jvm/im-domain/src/main/kotlin/im/domain/invitations/ChannelInvitation.kt
package im.domain.invitations

import im.domain.channel.Channel
import im.domain.channel.ChannelRole
import im.domain.user.User
import im.domain.wrappers.identifier.Identifier
import im.domain.wrappers.identifier.toIdentifier
import java.time.LocalDateTime

/**
 * Represents an invitation to a channel.
 *
 * @property id The unique identifier of the invitation.
 * @property channel The channel to which the user is invited.
 * @property inviter The user that sent the invitation.
 * @property invitee The user that received the invitation.
 * @property status The status of the invitation.
 * @property role The role that the user will have in the channel.
 * @property expiresAt The date and time when the invitation expires.
 */
data class ChannelInvitation(
    val id: Identifier = Identifier(0),
    val channel: Channel,
    val inviter: User,
    val invitee: User,
    val status: ChannelInvitationStatus = ChannelInvitationStatus.PENDING,
    val role: ChannelRole,
    val expiresAt: LocalDateTime,
) {
    companion object {
        operator fun invoke(
            id: Long = 0,
            channel: Channel,
            inviter: User,
            invitee: User,
            status: ChannelInvitationStatus = ChannelInvitationStatus.PENDING,
            role: ChannelRole,
            expiresAt: LocalDateTime,
        ): ChannelInvitation =
            ChannelInvitation(
                id.toIdentifier(),
                channel,
                inviter,
                invitee,
                status,
                role,
                expiresAt,
            )
    }

    val expired: Boolean
        get() = expiresAt.isBefore(LocalDateTime.now())

    val isValid: Boolean
        get() = !expired && status == ChannelInvitationStatus.PENDING

    init {
        require(channel.members.keys.contains(inviter)) { "Inviter must be a member of the channel" }
    }

    /**
     * Accepts the invitation.
     *
     * @return a new invitation with the status set to accepted
     */
    fun accept(): ChannelInvitation = copy(status = ChannelInvitationStatus.ACCEPTED)

    /**
     * Rejects the invitation.
     *
     * @return a new invitation with the status set to rejected
     */
    fun reject(): ChannelInvitation = copy(status = ChannelInvitationStatus.REJECTED)

    /**
     * Updates the role and expiration date of the invitation.
     */
    fun update(
        role: ChannelRole,
        expiresAt: LocalDateTime,
    ): ChannelInvitation = copy(role = role, expiresAt = expiresAt)
}

File: code/js/src/Domain/invitations/ChannelInvitation.ts
import { Channel, channelFromDto } from '../channel/Channel';
import { User, userFromDto } from '../user/User';
import { Identifier } from '../wrappers/identifier/Identifier';
import { ChannelInvitationStatus } from './ChannelInvitationStatus';
import { ChannelRole } from '../channel/ChannelRole';
import { ChannelInvitationCreationOutputModel } from '../../Dto/output/invitations/ChannelInvitationCreationOutputModel';
import { ChannelInvitationOutputModel } from '../../Dto/output/invitations/ChannelInvitationOutputModel';

/**
 * Represents a channel invitation.
 *
 * @param id The unique identifier of the channel invitation.
 * @param channel The channel to which the user is invited.
 * @param inviter The user who invited the invitee to the channel.
 * @param invitee The user who is invited to the channel.
 * @param status The status of the invitation.
 * @param role The role of the user in the channel.
 * @param expiresAt The date and time when the invitation expires.
 *
 */
export interface ChannelInvitation {
    id: Identifier;
    channel: Channel;
    inviter: User;
    invitee: User;
    status: ChannelInvitationStatus;
    role: ChannelRole;
    expiresAt: Date;
}

export function channelInvitationFromDto(
    dto: ChannelInvitationOutputModel,
): ChannelInvitation {
    return {
        id: new Identifier(dto.id),
        channel: channelFromDto(dto.channel),
        inviter: userFromDto(dto.inviter),
        invitee: userFromDto(dto.invitee),
        status: dto.status,
        role: dto.role,
        expiresAt: new Date(dto.expiresAt),
    };
}

export function channelInvitationFromCreation(
    dto: ChannelInvitationCreationOutputModel,
    channel: Channel,
    invitee: User,
    expiresAt: Date,
    role: ChannelRole,
): ChannelInvitation {
    return {
        id: new Identifier(dto.id),
        channel,
        inviter: channel.owner,
        invitee,
        status: ChannelInvitationStatus.PENDING,
        role,
        expiresAt,
    };
}
